pkg/server/server.go

```diff
@@ -434,6 +436,31 @@ func (s *Server) Run(ctx context.Context) error {
 			logger.Info("bootstrapping root workspace phase 0")
 			s.RootShardKcpClusterClient = s.KcpClusterClient
 
+			if s.Options.Extra.CoreIdentitiesFile != "" {
+				logger.Info("injecting core identities to root workspace", "file", s.Options.Extra.CoreIdentitiesFile)
+				logger.Info("bootstrapping system namespace into root workspace", "file", s.Options.Extra.CoreIdentitiesFile)
+				if err := configrootidentitysecretns.Bootstrap(hookCtx,
+					s.BootstrapApiExtensionsClusterClient.Cluster(core.RootCluster.Path()).Discovery(),
+					s.DynamicClusterClient.Cluster(core.RootCluster.Path()),
+					s.KubeClusterClient.Cluster(core.RootCluster.Path()),
+				); err != nil {
+					logger.Error(err, "failed to bootstrap identity secrets namespace")
+					return nil // don't klog.Fatal. This only happens when context is cancelled.
+				}
+				logger.Info("bootstrapped system namespace into root workspace", "file", s.Options.Extra.CoreIdentitiesFile)
+				logger.Info("injecting core identities into root workspace", "file", s.Options.Extra.CoreIdentitiesFile)
+				if err := configrootcoreidentities.Bootstrap(hookCtx,
+					s.BootstrapApiExtensionsClusterClient.Cluster(core.RootCluster.Path()).Discovery(),
+					s.DynamicClusterClient.Cluster(core.RootCluster.Path()),
+					s.KubeClusterClient.Cluster(core.RootCluster.Path()),
+					s.Options.Extra.CoreIdentitiesFile,
```

Comment:
E.g. here the file is read `pkg/server` and then the contents are handed off:
https://github.com/kcp-dev/kcp/blob/e2ba960b69964aa42e6923710f766fce62dc2024/pkg/server/config.go#L369-L389

Might also make it easier to test `config` without having to write the contents out to a file rather than just passing the bytes.

Reply:
Good point, reading in the caller function now.